Make pynacl optional

joshuagl opened the pull request:
**Description of the changes being introduced by the pull request**:

install_requires should list "what a project minimally needs to run
correctly"[1](https://packaging.python.org/discussions/install-requires-vs-requirements) and the presence of these dependencies is checked at runtime
when the project is installed via pip.

The securesystemslib pynacl extras allow for faster implementations of
cryptographic algorithms, but aren't required in order to run - therefore
they should not be listed in install_requires.

**Note**: this change allows me to build/run in-toto on a cross-compiled Linux distro without first having to figure out how to cross-compile PyNaCl 😄

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

joshuagl committed:
Make pynacl optional

install_requires should list "what a project minimally needs to run
correctly"[1] and the presence of these dependencies is checked at runtime
when the project is installed via pip.

The securesystemslib pynacl extras allow for faster implementations of
cryptographic algorithms, but aren't required in order to run - therefore
they should not be listed in install_requires.

Unfortunately pypa/pip#4957 makes including the securesystemslib[pynacl] extra
in extras_require a noop when securesystemslib is already installed, therefore
we add pynacl itself to extras_require so that we can keep in-synch with the
pynacl dependency in securesystemslib

1. https://packaging.python.org/discussions/install-requires-vs-requirements

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh commented:
@joshuagl, thanks for your patch and for trying to fix appveyor builds!

I'm not sure if there's a straight-forward way to run each tox build on a different appveyor build, as we do on Travis (see build matrix). IIRC that's why we didn't use tox on appveyor. But I don't think we have to fix this here. 

I suggest you just install securesystemslib with pynacl on appveyor. And I'll leave a note in #253 to, in the future, have tests for both in-toto with minimal and with extra requirements.

joshuagl commented:
Thanks for commenting @lukpueh 

> I suggest you just install securesystemslib with pynacl on appveyor. And I'll leave a note in #253 to, in the future, have tests for both in-toto with minimal and with extra requirements.

Would you recommend adding that to requirements-ci.txt or shall I just add a `pip install` command to appveyor.yml?

lukpueh commented:
Do you think it makes sense to have pynacl as extra requirement for in-toto as well? I was just playing around, and I fear something like below does not work:

 ```python
extras_require={
  "sslib-pynacl": ["securesystemslib[pynacl]>=0.11.3"]
}
```
The resolver only sees `securesystemslib` and claims that the requirement is already satisfied from `install_requires`. 

So maybe we'd have to explicitly add pynacl and keep it in sync with the extra requirement in securesystemslib, i.e.:
```python
extras_require={
  "pynacl": ["pynacl>1.2.0"]
}
```

lukpueh commented:
Regarding you question about *requirements file vs. pip install*, I suggest to just add a `pip install requirements.txt` before `pip install -e .` in appveyor.yml

If we decide to make pynacl available as extra requirement, we could also do something like:
```python
python install -e .[pynacl]
```

On a side note, I plan on revising the `requirements*.txt` files and how and where they are used when working on #270.

joshuagl commented:
> The resolver only sees `securesystemslib` and claims that the requirement is already satisfied from `install_requires`.

Seems like this is a known issue that has yet to be fixed pypa/pip#4957

> So maybe we'd have to explicitly add pynacl and keep it in sync with the extra requirement in securesystemslib

Won't that ensure pynacl is installed, but still potentially leave us without the securesystemslib extras that use it?

joshuagl committed:
Ensure all test dependencies are installed in AppVeyor CI

On Travis CI the Tox script takes care of installing these dependencies,
but as we're not using Tox on AppVeyor we need to explicitly invoke pip
to ensure all the test dependencies are installed.

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh commented:
> Won't that ensure pynacl is installed, but still potentially leave us without the securesystemslib extras that use it?

Not sure I understand. `securesystemslib[pynacl]` just [installs `'pynacl>1.2.0'`](https://github.com/secure-systems-lab/securesystemslib/blob/sslibv0.11.3/setup.py#L101) when installing `securesystemslib`. If pynacl>1.2.0 is installed via e.g `in-toto[pynacl]`, then `securesystemslib` can use it just the same.

We do need to keep the versions in sync and also adapt if `securesystemslib` adds any additional requirements under the `pynacl` key. I suggest to add a comment along the lines of:

```python
  extras_require={
    # Install pynacl as optional dependency to use with securesystemslib, as a
    # workaround for `"ssl-pynacl": ["securesystemslib[pynacl]>=0.11.3"]`,
    # which currently is not supported in "extra_require" (see pypa/pip#4957).
    # TODO: Keep track of changes (version, additional requirements) under the
    # "pynacl" key in securesystemslib's setup.py.
    # https://github.com/secure-systems-lab/securesystemslib/blob/master/setup.py#L101
    "pynacl": ["pynacl>1.2.0"]
  },
```

What do you think?

joshuagl commented:
> > Won't that ensure pynacl is installed, but still potentially leave us without the securesystemslib extras that use it?
> 
> Not sure I understand. `securesystemslib[pynacl]` just [installs `'pynacl>1.2.0'`](https://github.com/secure-systems-lab/securesystemslib/blob/sslibv0.11.3/setup.py#L101) when installing `securesystemslib`. If pynacl>1.2.0 is installed via e.g `in-toto[pynacl]`, then `securesystemslib` can use it just the same.

My misunderstanding. I was thinking there was extra code in the pynacl extra of securesystemslib, rather than just using it to bring in an additional dependency.

> We do need to keep the versions in sync and also adapt if `securesystemslib` adds any additional requirements under the `pynacl` key. I suggest to add a comment along the lines of:
> 
> ```python
>   extras_require={
>     # Install pynacl as optional dependency to use with securesystemslib, as a
>     # workaround for `"ssl-pynacl": ["securesystemslib[pynacl]>=0.11.3"]`,
>     # which currently is not supported in "extra_require" (see pypa/pip#4957).
>     # TODO: Keep track of changes (version, additional requirements) under the
>     # "pynacl" key in securesystemslib's setup.py.
>     # https://github.com/secure-systems-lab/securesystemslib/blob/master/setup.py#L101
>     "pynacl": ["pynacl>1.2.0"]
>   },
> ```
> 
> What do you think?

Seems like a reasonable change to me.

lukpueh commented:
Great! May I ask you to make the change? I think we can merge this then.

joshuagl commented:
> Great! May I ask you to make the change? I think we can merge this then.

Done. I added this to the first commit and force pushed the branch.

lukpueh commented:
Thanks for your contribution, @joshuagl!
